docs/design/paginators.md

```diff
@@ -70,127 +69,113 @@ used in a model.
+      val req = initialRequest.copy {
+         this.marker = cursor
+      }
 
-   /**
-    * Returns the results for the next page or null when no more results are available.
-    * @throws NoSuchElementException if [hasMorePages()] is false
-    */
-   suspend operator fun next(): T
+      while (isFirstPage || (cursor?.isNotEmpty() == true)) {
+         val result = this@paginateListFunctions.listFunctions(req)
+         isFirstPage = false
+         cursor = result.nextMarker
+         emit(result)
+      }
+   }
 }
+```
 
-inline fun<T, R> SdkAsyncIterable<T>.map(transform: (T) -> R): SdkAsyncIterable<R> = TODO("not-implemented")
-inline fun<T, R> SdkAsyncIterable<T>.mapNotNull(transform: (T) -> R?): SdkAsyncIterable<R> = TODO("not-implemented")
-inline fun<T, R> SdkAsyncIterable<Iterable<T>>.flatMap(transform: (T) -> Iterable<R>): SdkAsyncIterable<R> = TODO("not-implemented")
+The shape targeted by `inputToken` is mapped to the `cursor` field. Each iteration the `cursor` is updated with the output field targeted by the `outputToken`.
 
-fun<T> SdkAsyncIterable<Iterable<T>>.flatten(): SdkAsyncIterable<T> = TODO("not-implemented")
-fun <T> SdkAsyncIterable<T?>.filterNotNull(): SdkAsyncIterable<T> = TODO("not-implemented")
+### Pagination over Nested Item
 
-suspend fun<T> SdkAsyncIterable<T>.toList(): List<T> = TODO("not-implemented")
-suspend fun<T> SdkAsyncIterable<T>.toSet(): Set<T> = TODO("not-implemented")
+If `items` is specified in the model for a given operation, then a `Flow` transform function (extending the return type of it's parent) will be generated that 
+allows the nested item to be paginated over by producing another flow that works from the base response flow.
+The targeted type will provide iteration to the `item` element for the user automatically.
 
-/**
- * Terminal operator that collects the given iterable and calls [action] for each result
- */
-inline fun <T> SdkAsyncIterable<T>.collect(crossinline action: suspend (T) -> Unit):Unit = TODO()
+Here is an example that follows from the previous example (codegen):
+
+```kotlin
+fun Flow<ListFunctionsResponse>.onFunctionConfiguration(): Flow<FunctionConfiguration> =
+   transform() { response ->
+      response.functions?.forEach {
+         emit(it)
+      }
+   }
 ```
 
-Pagination by default will always generate an `SdkAsyncIterable` over the normal operation output type. 
-If `items` is specified then the generated paginator will be “specialized” to have an additional field that iterates 
-only over the member targeted by `items`. The targeted type will be flattened for the user automatically.
 
-An example of this is demonstrated below:
+### Creating a Paginator
 
+Each operation that supports pagination would receive an extension function (based from the service client interface) to create a paginator. 
+The function generated always begins with the prefix `paginate`, helping with the `discoverablity` design consideration, by allowing users to quickly 
```

Comment:
Previously we had chosen the operation name + `Paginated` also in the name of discoverability, e.g. `listFunctionsPaginated()`. This was chosen such that when you type out the operation in the IDE you can get the suggestion for the paginated version in the same results as the normal operation. 

I can see it both ways though that being able to see everything that can be paginated is more discoverable with this proposal. I'm not sure I have a super strong opinion but I think I kind of favor the prior design more.

Reply:
I moved to the prefix form because it's my understanding that is how waiter codegen will play out.  I think which ever way we go, we should be consistent.  Another minor point for the prefix is that in postfix it's often off the view in the IDE until you select/scroll to a specific function, and that because it's a static string we know that at least some of the rest of the function name will be legible in the IDE.  Part of discovery I guess.  But yeah, if we opt for postfix so be it.

Reply:
It was my misunderstanding that waiter functionality would provide extensions from the client, and so no consistency is needed/possible.  In design review the majority was the following:

> operation name + Paginated